qualtran/bloqs/comparison_gates.ipynb

```diff
@@ -0,0 +1,957 @@
+    "\n",
+    "    # Create the qubit that will represents Q.\n",
+    "    qubit_for_q = cirq.NamedQubit('q')\n",
+    "\n",
+    "    operations = []\n",
+    "\n",
+    "    q_0 = Q & 1  # The 0th bit of Q.\n",
+    "    q_1 = (Q >> 1) & 1  # The 1s bit of Q.\n",
+    "\n",
+    "    # Copy q_0 to `qubit_for_q`\n",
+    "    if q_0:\n",
+    "        operations.append(cirq.X(qubit_for_q))\n",
+    "\n",
+    "    # Store the xor of p_1 and q_1 into ancilla\n",
+    "    operations.append(cirq.CNOT(P[1], ancilla))\n",
+    "    if q_1:\n",
+    "        operations.append(cirq.X(ancilla))\n",
+    "\n",
+    "    # Do CSWAP on `P`\n",
+    "    operations.append(cirq.CSWAP(ancilla, *P))\n",
+    "\n",
+    "    # Second CSWAP collapses to either identity or CNOT.\n",
+    "    if q_0 != q_1:\n",
+    "        operations.append(cirq.CNOT(ancilla, qubit_for_q))\n",
+    "\n",
+    "    return operations, qubit_for_q"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# For example when Q = 2 this is what the circuit looks like.\n",
+    "circuit, q = quantum_classical_mixer(first_quantum_number, 2)\n",
+    "circuit = cirq.Circuit(circuit)\n",
+    "circuit"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Now lets add SingleQubitCompare and built the full comparator.\n",
+    "def quantum_quantum_comparator_logn(\n",
+    "    P: Sequence[cirq.Qid], Q: int, z: cirq.Qid, less_than: cirq.Qid, greater_than: cirq.Qid\n",
+    "):\n",
+    "    mixer, q = quantum_classical_mixer(P, Q)\n",
+    "    yield mixer\n",
+    "    anc = cirq.NamedQubit.range(2, prefix='anc')\n",
+    "    yield cirq_ft.algos.SingleQubitCompare()(P[1], q, *anc)\n",
+    "\n",
+    "    yield cirq.CNOT(anc[0], less_than)\n",
+    "    yield cirq.CNOT(q, z)\n",
+    "    yield cirq.CNOT(anc[-1], greater_than)\n",
+    "\n",
+    "    yield cirq_ft.algos.SingleQubitCompare(adjoint=True)(P[1], q, *anc)\n",
+    "    yield from reversed([op**-1 for op in mixer])\n",
+    "\n",
+    "\n",
+    "q_n = 2\n",
+    "circuit = cirq.Circuit(\n",
+    "    quantum_quantum_comparator_logn(first_quantum_number, q_n, z, less_than, greater_than)\n",
+    ")\n",
+    "circuit"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Now we check individual possiblities.\n",
+    "check_each_possibility(circuit, first_quantum_number, [less_than, z, greater_than], qn=q_n)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "check_uniform_superposition(circuit, first_quantum_number, [less_than, z, greater_than])"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### T Complexity Analysis\n",
```

Comment:
You do a lot of verification of the behavior, which is good, but no verification of the T count. Maybe add a method that counts T gates and run it on the circuits?

Reply:
added